[FLINK-2837][storm] various improvements for Storm compatibility

mxm opened the pull request:
This pull request contains various fixes. Most prominently, the parsing logic for Storm topologies has been changed to support multiple inputs. The API has been redone slightly. Two new examples have been added.

- refactor to use Storm's topology builder

- remove FlinkTopologyBuilder

- instantiate context-based StreamExecutionEnvironment (local or remote)

- remove Flink and Storm behavior replicating classes

- modify FlinkTopology to parse Storm topology directly

- replace StormTestBase with StreamingTestBase

- let the FiniteFileSpout finish in corner cases

- expose taskId, fix off by one task id

- add print example

- FlinkTopologyBuilder changes (check if all inputs are available before processing)

- correct package typo

- support getter methods on TupleWrapper

- two input support

- add join example

- update docs

- use Flink file system access

mjsax commented on docs/apis/storm_compatibility.md:
```diff
@@ -57,20 +57,22 @@ See *WordCount Storm* within `flink-storm-examples/pom.xml` for an example how t
+`nimbus.host` and `nimbus.thrift.port` are used as `jobmanger.rpc.address` and
+`jobmanger.rpc.port`, respectively.  If a parameter is not specified, the value
+is taken from `flink-conf.yaml`.
 
```
Is there any change here? Looks like reformatting? I used a "single-line per sentence" formatting to make reviewing changes simples. Please stick to this formatting.

mxm replied:
Minor typo "can be uses" -> "can be used". I think fixed line length is the best because sentences can span multiple lines. If it is an issue I can revert it.

mjsax replied:
I like the single-line format because it make reviewing easier. If you have fixed line length, changing a small thing can lead for reformatting of a whole paragraph due to new line breaks etc. This make spotting the actual change quite hard. (I would appreciate it, if we could keep the current formatting -- but if is not "an issue" in the strong sense).

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/util/BoltFileSink.java:
```diff
@@ -18,20 +18,23 @@
 
-	private final String path;
+	private final Path path;
 	private BufferedWriter writer;
```
Please do not use `Path` -- **all** spouts/bolts should be written the "Storm" way and not include dependencies to Flink.

mxm replied:
Hmm but they depend on Flink anyways because of the Maven dependency?

mjsax replied:
That's not the point. If I re-use an existing bolt, I don't want to change anything. And I want to be able (from an example point of view) to run this bolt as-is in Storm, too. (Without any dependencies to Flink)

mxm replied:
So people download Flink and learn about the Storm compatibility layer to write spouts/bolts which they use in Storm topologies that run without Flink? That is beyond my imagination :)

mjsax replied:
No. People will have spout/bolt code they do not want to touch when running it in Flink. Thus there code will be written in the same way -- and so should be example be. Otherwise, we deliver the impression they need to change there code -- but they don't. Thus, we implement the example Spout/Bolts in a pure Storm way. Of course, if somebody develops a new Spout/Bolt with Flink in mind, your approach makes sense. However, this is not the main focus (it would be even better if the code new stuff Flink native in embedded mode, instead of develop Spout/Bolts which are Flink tailored).

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/util/BoltFileSink.java:
```diff
@@ -40,16 +43,17 @@ public BoltFileSink(final String path) {
-			this.writer = new BufferedWriter(new FileWriter(this.path));
+			FSDataOutputStream outputStream = FileSystem.getLocalFileSystem().create(path, false);
+			this.writer = new BufferedWriter(new OutputStreamWriter(outputStream));
 		} catch (final IOException e) {
```
Same here.

mxm replied:
I had problems if the path contained a file scheme without the file system abstraction.

mjsax replied:
In my examples (ITCases) I strip the `file://" prefix before given the path to the spout.

mxm replied:
That's odd but if so I'd change this to strip the file:// prefix.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/util/FileSpout.java:
```diff
@@ -38,6 +38,8 @@
 
+	protected boolean finished;
+
 	public FileSpout() {}
```
`FileSpout` should be implemented the Storm way -- to set the "finished"
flag here does not make sense from a Storm point of view (there is no
such thing as a finite spout)

mxm replied:
But this doesn't alter the behavior of FileSpout. It is still infinite.

mjsax replied:
Yes. But as the argument from above explains. All spout/bolts should be written in a Flink agnostic way. No Storm developer would write it like this, because this flag does only make sense if you know about `FinitSpout` interface -- but you should not assume to know about it.

mxm replied:
Okay, I'll revert although `FiniteSpout` is Flink specific as well...

mjsax replied:
The point it the following. `FileSpout` is a Flink agnostic implementation that is improved in a Flink aware way by `FiniteFileSpout`. Thus, `FileSpout` should be implemented the Storm way without any knowledge of Flink. And `FiniteFileSpout` should use Flink stuff, but not change the code of `FileSpout`.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/util/FiniteFileSpout.java:
```diff
@@ -32,46 +23,17 @@
-		}
-		return line == null;
+		return finished;
 	}
```
This example shows how a regular Storm spout can be improved
using FiniteSpout interface -- I would keep it as is (even if seems to
be unnecessary complicated -- imagine that you don't have the code of
FileSpout)

mxm replied:
I don't understand why I should copy code instead. It is more elegant this way.

mjsax replied:
It is more elegant. But you have already Flink in mind. Given the Flink agnostic `FileSpout` we have to do it this way. (assume you do not have the source code of `FileSpout` but only the class file and want to extend it with `FiniteSpout` interface)

mxm replied:
Following your argument, we should also remove the `FiniteSpout` functionality, because it is not included in Storm as well. I'll revert this but I don't understand your argument.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkClient.java:
```diff
@@ -183,10 +183,10 @@ public void submitTopologyWithOpts(final String name, final String uploadedJarLo
 		if (this.conf != null) {
-			topology.getConfig().setGlobalJobParameters(new StormConfig(this.conf));
+			topology.getExecutionEnvironment().getConfig().setGlobalJobParameters(new StormConfig(this.conf));
 		}
```
Why this change? Just want to understand?

mxm replied:
Because FlinkTopology is not subclassed from StreamExecutionEnvironment anymore. The class has been removed and FlinkTopologyBuilder is now FlinkTopology.

mjsax replied:
I see. Makes sense now.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkClient.java:
```diff
@@ -183,10 +183,10 @@ public void submitTopologyWithOpts(final String name, final String uploadedJarLo
 
-		final StreamGraph streamGraph = topology.getStreamGraph();
+		final StreamGraph streamGraph = topology.getExecutionEnvironment().getStreamGraph();
 		streamGraph.setJobName(name);
```
Same here? Why changed?

mxm replied:
Because FlinkTopology is not subclassed from StreamExecutionEnvironment anymore.

mjsax replied:
I see. Makes sense now.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkLocalCluster.java:
```diff
@@ -48,12 +49,10 @@
 	public FlinkLocalCluster() {
-		this.flink = new LocalFlinkMiniCluster(new Configuration(), true, StreamingMode.STREAMING);
-		this.flink.start();
 	}
```
What is the advantage by removing this?

mxm replied:
Lazy startup of the flink cluster. This enables us to set the right number of task slots for the job.

mjsax replied:
I just realized that I have something similar in my current work on FLINK-2721 ;)

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java:
```diff
@@ -15,75 +16,468 @@
+										GlobalStreamId streamId, Grouping grouping,
+										Map<String, DataStream<Tuple>> producer) {
+
+		Preconditions.checkNotNull(userBolt);
```
I would use `assert` here because it a private method.

mxm replied:
Asserts are not enabled by default. I think it's important to always check this condition.

mjsax replied:
It's a matter of philosophy I guess. I think it is good enough to test via `assert` (which are enabled in the test). No external user can call this method, so checking if the calling code is correct in the test is sufficient to me.

mxm replied:
Yes, but these checks can be executed during run time of the user program. Assert only checks these when debugging is enabled.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java:
```diff
@@ -15,75 +16,468 @@
+														GlobalStreamId streamId, DataStream<Tuple> inputStream,
+														GlobalStreamId streamId2, DataStream<Tuple> inputStream2) {
+		Preconditions.checkNotNull(boltId);
+		Preconditions.checkNotNull(streamId);
```
`assert`

mxm replied:
Asserts are not enabled by default. I think it's important to always check this condition.

mjsax replied:
See above.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -75,11 +78,13 @@
-	public BoltWrapper(final IRichBolt bolt) throws IllegalArgumentException {
-		this(bolt, null, (Collection<String>) null);
+	public BoltWrapper(final IRichBolt bolt, String componentId, String streamId) throws IllegalArgumentException {
+		this(bolt, componentId, streamId, null, (Collection<String>) null);
```
This makes the interface quite hard to use and should not be necessary. The `TopologyContext` created in `open()` contains this information.

mxm replied:
The naming is a bit odd but this is actually the component and stream id of the *producer*.

mjsax replied:
`TopologyContext` contains this information, too.

mxm replied:
Okay, I will check if I can use the information contained in TopologyContext.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/StormTuple.java:
```diff
@@ -44,6 +45,21 @@
+	private final String streamId;
+	private final MessageId id;
+	private final String componentId;
+
```
There is a JIRA for this. https://issues.apache.org/jira/browse/FLINK-2721 I am already working on it. I think that FLINK-2721 should be resolved before this.

mxm replied:
This is fixed in this PR.

mjsax replied:
I have already worked on this and I think my solution is smoother as it compute all this stuff automatically under the hood (without the need that the user specifies the name in the constructor). No PR yet as not finished completely. Need to get in sync about it.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/WrapperSetupHelper.java:
```diff
@@ -150,7 +153,7 @@ static synchronized TopologyContext createTopologyContext(
 
-			taskId = context.getIndexOfThisSubtask();
+			taskId = context.getIndexOfThisSubtask() + 1;
 
```
Why `+1` ?

mxm replied:
Because Flink's task index starts with 0 but Storm starts with 1...

mjsax replied:
Are you sure about this? I doubt it (but not 100% sure)

mxm replied:
Actually, it doesn't matter. I set this before changing the topology parsing logic. For some topologies it would only run with this fix. But this has been fixed so the +1 is not necessary anymore.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/WrapperSetupHelper.java:
```diff
@@ -187,14 +190,15 @@ static synchronized TopologyContext createTopologyContext(
-			assert (taskId != null);
+
+			Preconditions.checkNotNull("Task ID may not be null!", taskId);
 		}
```
I prefer `assert` for internal method.

mxm replied:
They are disabled by default and are not reliable enough.

mjsax replied:
Why not readable enough?

mxm replied:
Not "reliable enough" because they are disabled by default :)

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/wrappers/StormTupleTest.java:
```diff
@@ -595,7 +593,7 @@ public void testGetBinaryByFieldPojoGetter() throws Exception {
 
-		assert (index < arity);
+		Assert.assertTrue(index < arity);
 
```
This is not a assertion for a test result; thus usage of `assert` is correct (checks if test is written properly).

mxm replied:
Hmm, it's part of the test, no? Doesn't really matter but in general I thought usage of assert is discouraged.

mjsax replied:
Is it? I am not aware if this. Why? Using `assert` to check internal code invariants is the best way to go IMHO.

mxm replied:
Because they are disabled during run time. This makes bugs e.g. in stack traces harder to find.

mjsax replied:
But not in testing scenario (ie, while executing JUnit and ITCase asserts are enabled).

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/wrappers/WrapperSetupHelperTest.java:
```diff
@@ -180,8 +178,6 @@ public void testCreateTopologyContext() {
 				.shuffleGrouping("bolt1", TestDummyBolt.groupingStreamId)
-				.shuffleGrouping("bolt1", TestDummyBolt.shuffleStreamId)
-				.shuffleGrouping("bolt2", TestDummyBolt.groupingStreamId)
 				.shuffleGrouping("bolt2", TestDummyBolt.shuffleStreamId);
```
Why did you change this?

mxm replied:
Because only number of inputs <= 2 are supported by the translator :)

mjsax replied:
We need to fix this... Supporting only 2 inputs is quite weak. Think about a union of many input streams... Join is not the only use case for multiple inputs.

mxm replied:
Agreed. My concern was to get the join working but union should also be supported for same data output types. Could we do that in a follow-up?

mjsax replied:
Well. From a Storm point of view, there is only `union`. As it is the generic Storm case it includes the join case. I guess you specialized join solution would be obsolete after generic union is supported. Therefore, I would prefer to get it right from the beginning on... My idea would be to try to get rid of `TwoInputBoltWrapper` and "union" the incoming streams somehow to feed a single stream to `BoltWrapper`. The tricky part is, that we cannot use Flink's `union` because it assume the same input type, but Storm can union different types into one stream... What do you think about this idea?

mxm replied:
Well. Get it right from the beginning. I think it was all but right until now :) And in this regard, its much more defined now. At least you get an error if you have more than two inputs.

I agree that we should fix this. But it's going to be a bit tricky because we have to hack around Flink's limitation. I would rather not do this in this pull request.

mjsax replied:
I see you point. But this PR might be too big anyway. You try to do 3 thing at the same time (two are backed up by a JIRA). How hard would it be to split this PR? Last but not least, the multi-input-stream JIRA is not resolved by this. [And the second JIRA you try to resolve is assigned to me, and I have already worked on it -- I actually would like to finish my work on it]

mxm replied:
I was actually not specifically trying to address JIRA issues but just fixed everything I discovered on the way while trying out the compatibility layer. Only after fixing I realized there are open JIRA issues. One is assigned to me (FLINK-2837] and the other one (FLINK-2721) is open since two months. I think it would be a shame not to merge this pull request soon. It provides a good foundation to address any further issues. Splitting this PR should not be trivial with all the changes.

I already accommodated you with the API changes. Also, I would like to address most of your comments but I'm not too inclined to split up this PR (if it is even possible). Could you base your work on this pull request and do a follow-up?

mjsax replied:
Don't understand me wrong. I don't want to discard your work! And I believe that you did not intent do get a "messy" PR. But that's the current state.

I think we can refine and merge it. But it does not resolve FLINK-2837 even if it improves on it. I would also assume, that your union code will be reworked heavily later on... Not sure about your tuple meta information code. Need to have a look in detail. That is the reason why I had the idea to apply the discussed API changes only in a single PR. But if this is too complex, we should just carry on with this PR.

Btw: even if the JIRA is quite old it is not assigned to you; thus you should have ask about it. You did the same with FLINK-2837 which was assigned to me, too -- I did not work in it yet so a assigned it to you (I thought as you did have the union code together with the API changes, that should be fine).

Additionally, the reason I just assigned it to you was, that FLINK-2837 is actually a requirement for FLINK-2721. That is why I stopped working on it back than, but did not have time to fix FLINK-2837 either. I did not assume that you tackle the join-case which does require the tuple meta info... A regular union does not require it.

Anyway. Just let us get this PR done. :)

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java:
```diff
@@ -15,75 +16,468 @@
+					default:
+						throw new UnsupportedOperationException("Don't know how to translate a bolt "
+								+ boltId + " with " + numberOfInputs + " inputs.");
+				}
```
We should try to fix this...

mxm replied:
This is an improvement. Before, it would just blindly take one input at a time and create an operator. As far as I understand, Flink only supports two inputs at a time. I would like to address any further improvements in this regard after the merge.

mxm commented:
I will integrate the changes of #1387. It would be great if we merged these changes afterwards.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/join/SingleJoinExample.java:
```diff
@@ -0,0 +1,86 @@
+	public static void main(String[] args) throws Exception {
+		final FeederSpout genderSpout = new FeederSpout(new Fields("id", "gender"));
+		final FeederSpout ageSpout = new FeederSpout(new Fields("id", "age"));
+
```
Can you extend this to use different number of attributes. Makes test more generic I think.

mxm replied:
You mean different number of fields?

mjsax replied:
Yes. :) For example, 2 fields for one input and 3 fields for the other input. (The idea is to have different number for both inputs to make sure the input schemas of both can differ not only by their types and also by their number of attributes)

mxm replied:
Okay should be doable.

mxm commented:
I've rebased to the latest master and addressed your comments. I would like to merge this and programmatically fix the multiple inputs issue afterwards.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/print/PrintSampleStream.java:
```diff
@@ -0,0 +1,61 @@
+ * Prints incoming tweets. Tweets can be filtered by keywords.
+ */
+public class PrintSampleStream {        
+	public static void main(String[] args) throws Exception {
```
What is the purpose of this example? Does it show anything special?

mxm replied:
It shows how to run an existing Storm topology with Flink. It prints from Twitter which is kind of neat. It's also included in Storm. It's nice to have some other examples other than WordCount. This was actually not working before this PR...

mjsax replied:
I cannot see why it did not work before? Can you explain what the problem was?

mxm replied:
The problem was that the `BoltWrapper` wouldn't create a `BoltCollector` if the bolt didn't define any output fields. That led to a NullPointerException.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/split/operators/RandomSpout.java:
```diff
@@ -17,9 +17,6 @@
 
-import java.util.Map;
-import java.util.Random;
-
```
pure reformatting

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/split/operators/VerifyAndEnrichBolt.java:
```diff
@@ -17,8 +17,6 @@
 package org.apache.flink.storm.split.operators;
 
-import java.util.Map;
-
```
pure reformatting

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/wordcount/WordCountLocal.java:
```diff
@@ -57,16 +57,13 @@ public static void main(final String[] args) throws Exception {
 		// build Topology the Storm way
-		final FlinkTopologyBuilder builder = WordCountTopology.buildTopology();
+		final TopologyBuilder builder = WordCountTopology.buildTopology(false);
 
```
Please remove `false` -- this test should use index (and not name) to specify the key. `WordCountLocalByName` does it the other way round such that both cases are covered.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/wordcount/SpoutSourceWordCount.java:
```diff
@@ -19,7 +19,6 @@
 
 import backtype.storm.topology.IRichSpout;
 import backtype.storm.utils.Utils;
-
```
pure reformatting

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/wordcount/operators/WordCountDataPojos.java:
```diff
@@ -17,10 +17,10 @@
 
-import java.io.Serializable;
-
 import org.apache.flink.examples.java.wordcount.util.WordCountData;
```
pure reformatting

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/wordcount/operators/WordCountFileSpout.java:
```diff
@@ -17,10 +17,9 @@
 package org.apache.flink.storm.wordcount.operators;
 
-import org.apache.flink.storm.util.FileSpout;
-
```
pure reformatting

mjsax commented on flink-contrib/flink-storm-examples/src/test/java/org/apache/flink/storm/split/SplitBolt.java:
```diff
@@ -17,8 +17,6 @@
 package org.apache.flink.storm.split;
 
-import java.util.Map;
-
```
pure reformatting

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkLocalCluster.java:
```diff
@@ -99,6 +111,7 @@ public void rebalance(final String name, final RebalanceOptions options) {
 
 	public void shutdown() {
 		flink.stop();
+		flink = null;
```
Should be kept. Otherwise, calling `submitTopologyWithOpts` a second time will run into a NPE. (Or add proper exception as `else` of `if (flink == null)` check, ie, "Cannot run topology. Cluster got shut down." or similar.)

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkOutputFieldsDeclarer.java:
```diff
@@ -20,11 +20,9 @@
 import backtype.storm.topology.OutputFieldsDeclarer;
 import backtype.storm.tuple.Fields;
 import backtype.storm.utils.Utils;
-
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java:
```diff
@@ -15,75 +16,474 @@
-	 * 		at every invocation
+	 * Creates a Flink program that uses the specified spouts and bolts.
+	 * @param stormBuilder The storm topology builder to use for creating the Flink topology.
+	 * @return A Flink Topology which may be executed.
```
typo: Strom with capital letter
"Flink program" or "Flink job" (not "Flink topology" -- if you keep it as is also fine. -- I personally prefer the use topology only for Storm to avoid confusion in terminology)
Typo: topology with lower case letter

mxm replied:
Let me get this straight. You're saying you don't want to use the name Flink topology but you are the author of the classes called FlinkTopology and FlinkTopologyBuilder (deleted by now). This is still a user-facing documentation and there we wanted to maintain the Storm vocabulary, right?

mjsax replied:
Sorry for being not consistent by myself -- I always try hard (but not always with success). The name `FlinkTopology` is kind of hybrid as it brides between both systems. I would claim that a `FlinkTopology` is a Flink streaming program that was derived from a Storm topology. And as a stated above: **keep it as is also fine**

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java:
```diff
@@ -15,75 +16,474 @@
-	 *
-	 * @throws UnsupportedOperationException
-	 * 		at every invocation
+	 * Returns the underlying Flink ExecutionEnvironment for the Storm topology.
```
add JavaDoc link to `ExecutionEnvironment`

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/api/FlinkTopology.java:
```diff
@@ -15,75 +16,474 @@
+	 * remote when executed thorugh ./bin/flink).
+	 * @return The execution result
+	 * @throws Exception
 	 */
```
Can we describe the exception in a meaningful way? Or are there too many reasons to get listed here? What does JavaDoc of `env.execute()` say?

mxm replied:
Yes, execute just throws an Exception. Nothing to explain here. StreamExecutionEnvironment says
> 	 * @throws Exception which occurs during job execution.

mjsax replied:
Maybe just put something similar -- even if it's not too useful... I thinks leaving it blank is even worse... (and if we really activate stricter code style checks once -- if this ever happens ;) -- it is already complete.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltCollector.java:
```diff
@@ -19,7 +19,6 @@
 import backtype.storm.task.IOutputCollector;
 import backtype.storm.tuple.Tuple;
-
 import org.apache.flink.api.java.tuple.Tuple0;
```
pure reformatting

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -53,21 +51,26 @@
+
+	protected final String inputComponentId;
+	protected final String inputStreamId;
+
```
Please add JavaDoc to those three.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -77,11 +80,13 @@
 	 * @param bolt
 	 *            The Storm {@link IRichBolt bolt} to be used.
+	 * @param inputStreamId
+	 * @param inputComponentId
```
JavaDoc incomplete -- same below -- will not mark it again. Please complete everywhere.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -89,17 +94,16 @@ public BoltWrapper(final IRichBolt bolt) throws IllegalArgumentException {
-	 * 
-	 * @param bolt
+	 *  @param bolt
 	 *            The Storm {@link IRichBolt bolt} to be used.
```
delete one space before `@param`

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -89,17 +94,16 @@ public BoltWrapper(final IRichBolt bolt) throws IllegalArgumentException {
 	 * @param inputSchema
-	 *            The schema (ie, ordered field names) of the input stream.
-	 * @throws IllegalArgumentException
-	 *             If the number of declared output attributes is not with range [0;25].
```
Why do you delete `@throws` ? More documentation is always better.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -108,20 +112,19 @@ public BoltWrapper(final IRichBolt bolt, final Fields inputSchema)
 	 * of {@link Tuple0} to {@link Tuple25} depending on the bolt's declared number of attributes.
-	 * 
-	 * @param bolt
+	 *  @param bolt
```
space

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -108,20 +112,19 @@ public BoltWrapper(final IRichBolt bolt, final Fields inputSchema)
 	 * @param rawOutputs
 	 *            Contains stream names if a single attribute output stream, should not be of type {@link Tuple1} but be
-	 *            of a raw type.
-	 * @throws IllegalArgumentException
```
keep `@throws`

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapperTwoInput.java:
```diff
@@ -0,0 +1,130 @@
+
+	private final String componentId2;
+	private final String streamId2;
+
```
missing JavaDoc for both members

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapperTwoInput.java:
```diff
@@ -0,0 +1,130 @@
+*            The schema (ie, ordered field names) of the input stream.    @throws IllegalArgumentException
+*             If {@code rawOuput} is {@code true} and the number of declared output attributes is not 1 or if
+*             {@code rawOuput} is {@code false} and the number of declared output attributes is not with range
+	 * */
```
formatting (space and stars) incomplete JavaDoc; missing `@throws`

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/FlinkTopologyContext.java:
```diff
@@ -27,13 +27,12 @@
 import backtype.storm.task.TopologyContext;
 import backtype.storm.tuple.Fields;
+import clojure.lang.Atom;
 
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax replied:
I see. DataArtians committer can do any change, but external committers get bullied if they apply similar changes... It is not against you or the change itself -- it unifies the style which does make sense. But I got bullied multiple times in other PRs when I did similar stuff...

mxm replied:
Not sure who is bullying whom :) Look at the classes and you will see that all imports are arranged like this. We want to be consistent, right? According to your suggestion, I changed the other import statements which were just reformatting.

Open source is often about compromises. Very rarely you will find that the code style of a person reflects exactly how you would do it. I'm making compromises and changing things as you like them. That's fine for me. Please don't give me a harder time by blaming my employer. I'm not aware I have done something like this to you. Next time you get blamed for something like this, please contact me and I'll try to help you. I don't think this is the right place to sort out things.

mjsax replied:
Good we are on the same page. And I don't want to bully you! I just mentioned the classes that do not contain any actual code change -- actually, according to the coding guidelines -- there should be no import-order changes even in the classes with code changes -- I did not comment on them -- just on the classes with pure reformatting. I like consistency so please apply the changes to all classes. But I did import-reorderings or making code formatting consistent (if it was inconsistent) and was always told "don't do this". So if it is a general rule, I just point it out here, too. I did not come up with the rule. And I never force my own code style -- a always adapt to the given style. :) It's is really about time to get a proper maven formatting tool running to get rid of all this stupid discussions. (And a said already: "It is not against you or the change itself" -- but the process seems to be inconsistent -- people follow the rules more or less strictly)

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/SetupOutputFieldsDeclarer.java:
```diff
@@ -17,12 +17,12 @@
 package org.apache.flink.storm.wrappers;
 
-import java.util.HashMap;
-
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/SpoutCollector.java:
```diff
@@ -18,7 +18,6 @@
 
 import backtype.storm.spout.ISpoutOutputCollector;
-
 import org.apache.flink.api.java.tuple.Tuple0;
```
pure reformatting

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/SpoutWrapper.java:
```diff
@@ -33,7 +30,8 @@
-import com.google.common.collect.Sets;
+import java.util.Collection;
+import java.util.HashMap;
 
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/StormTuple.java:
```diff
@@ -44,16 +45,30 @@
 	 * @param schema
-	 * 		The schema (ie, ordered field names) of the tuple.
+	 * @param producerComponentId
 	 */
```
formatting; incomplete.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/api/FlinkOutputFieldsDeclarerTest.java:
```diff
@@ -18,9 +18,7 @@
 import backtype.storm.utils.Utils;
-
 import org.apache.flink.api.common.typeinfo.TypeInformation;
-import org.apache.flink.storm.api.FlinkOutputFieldsDeclarer;
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/api/FlinkTopologyTest.java:
```diff
@@ -14,50 +14,70 @@
 public class FlinkTopologyTest {
 
-	@Test
-	public void testDefaultParallelism() {
```
Why removing this test?

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/api/FlinkTopologyTest.java:
```diff
@@ -14,50 +14,70 @@
 	@Test
-	public void testNumberOfTasks() {
-		final FlinkTopology topology = new FlinkTopology();
+	@Ignore
```
Please enable this test. I forgot to do this in my last commit which fixes this issue...

mxm replied:
ok

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/api/FlinkTopologyTest.java:
```diff
@@ -14,50 +14,70 @@
-		topology.increaseNumberOfTasks(2);
-		Assert.assertEquals(5, topology.getNumberOfTasks());
+	@Test
+	@Ignore
```
Please enable this test. I forgot to do this in my last commit which fixes this issue...

mxm replied:
ok

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/api/TestBolt.java:
```diff
@@ -16,14 +16,14 @@
 
+import java.util.Map;
+
 public class TestBolt implements IRichBolt {
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/api/TestSpout.java:
```diff
@@ -16,13 +16,13 @@
+import java.util.Map;
+
 public class TestSpout implements IRichSpout {
 	private static final long serialVersionUID = -4884029383198924007L;
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/util/TestDummyBolt.java:
```diff
@@ -26,6 +24,8 @@
 
+import java.util.Map;
+
 public class TestDummyBolt implements IRichBolt {
```
pure reformatiing

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/util/TestDummySpout.java:
```diff
@@ -26,6 +24,8 @@
 
+import java.util.Map;
+
 public class TestDummySpout implements IRichSpout {
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/util/TestSink.java:
```diff
@@ -16,16 +16,16 @@
+import java.util.List;
+import java.util.Map;
+
 public class TestSink implements IRichBolt {
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/wrappers/SpoutWrapperTest.java:
```diff
@@ -21,7 +21,6 @@
 import backtype.storm.topology.IRichSpout;
 import backtype.storm.tuple.Fields;
-
 import org.apache.flink.api.common.ExecutionConfig;
```
pure reformatting

mxm replied:
It follows the import style of the other classes, so I'll leave this as it is.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/StormTuple.java:
```diff
@@ -44,16 +45,30 @@
+	 */
+	StormTuple(final IN flinkTuple, final Fields schema) {
+		this(flinkTuple, schema, -1, "testStream", "componentID");
+	}
```
Are this meaningful/helpful defaults? Why not just set it to `null`?

mxm replied:
The use of null is often problematic. I prefer default values.

mjsax replied:
Ok. I guess it would make sense to use Storm's `Utils.DEFAULT_STREAM_ID` here? And maybe add a `public final static String DEFAULT_OPERATOR_ID` variable to `StormTuple`? What about using "defaultID" or "unspecified" instead of "componentID" or similar? Just to make it clear if the name shows up in the UI?

mxm replied:
Fair enough, I use default id constants now.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/wrappers/StormTupleTest.java:
```diff
@@ -613,29 +611,29 @@ public void testGetBinaryByFieldPojoGetter() throws Exception {
+	@Test
 	public void testGetSourceGlobalStreamid() {
-		new StormTuple<Object>(null, null).getSourceGlobalStreamid();
+		Assert.assertNotNull(new StormTuple<Object>(null, null).getSourceGlobalStreamid());
```
Can we improve on all this tests? Just a check for "not-null" seems a  little limited to me. We should rather check for defaults values and also check the full constructor case.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/exclamation/ExclamationWithBolt.java:
```diff
@@ -72,7 +71,7 @@ public static void main(final String[] args) throws Exception {
 						TypeExtractor.getForObject(""),
 						new BoltWrapper<String, String>(new ExclamationBolt(),
-								new String[] { Utils.DEFAULT_STREAM_ID }))
+								"stream", "component", new String[] { Utils.DEFAULT_STREAM_ID }))
```
Can't we use default values for streamID and componentID here, ie, obit both parameters?

mxm replied:
Can do.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/split/SpoutSplitExample.java:
```diff
@@ -70,7 +70,7 @@ public static void main(final String[] args) throws Exception {
 				TypeExtractor.getForObject(new Tuple2<String, Integer>("", 0)),
 				new BoltWrapper<SplitStreamType<Integer>, Tuple2<String, Integer>>(
-						new VerifyAndEnrichBolt(false)))
+						new VerifyAndEnrichBolt(false), "stream", "component"))
```
Default values for both? Or use `ODD_STREAM` for consistency.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/wordcount/BoltTokenizerWordCount.java:
```diff
@@ -64,7 +63,7 @@ public static void main(final String[] args) throws Exception {
 				.transform("BoltTokenizer",
 						TypeExtractor.getForObject(new Tuple2<String, Integer>("", 0)),
-						new BoltWrapper<String, Tuple2<String, Integer>>(new BoltTokenizer()))
+						new BoltWrapper<String, Tuple2<String, Integer>>(new BoltTokenizer(), "stream", "component"))
```
Same question here.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/wordcount/BoltTokenizerWordCountPojo.java:
```diff
@@ -71,7 +70,7 @@ public static void main(final String[] args) throws Exception {
 				.transform("BoltTokenizerPojo",
 						TypeExtractor.getForObject(new Tuple2<String, Integer>("", 0)),
-						new BoltWrapper<Sentence, Tuple2<String, Integer>>(new BoltTokenizerByName()))
+						new BoltWrapper<Sentence, Tuple2<String, Integer>>(new BoltTokenizerByName(), "stream", "component"))
```
Again.

mjsax commented on flink-contrib/flink-storm-examples/src/main/java/org/apache/flink/storm/wordcount/BoltTokenizerWordCountWithNames.java:
```diff
@@ -75,7 +74,7 @@ public static void main(final String[] args) throws Exception {
 						TypeExtractor.getForObject(new Tuple2<String, Integer>("", 0)),
 						new BoltWrapper<Tuple1<String>, Tuple2<String, Integer>>(
-								new BoltTokenizerByName(), new Fields("sentence")))
+								new BoltTokenizerByName(), "stream", "component", new Fields("sentence")))
```
Again.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/wrappers/WrapperSetupHelperTest.java:
```diff
@@ -193,24 +189,22 @@ public void testCreateTopologyContext() {
 			cluster.shutdown();
 
-			if (TestSink.result.size() == 8) {
+			if (TestSink.result.size() >= 4) {
```
Why `>=` and not `==` ?

mxm replied:
The Storm executor sometimes returned more results for me. I've adjusted it to a fixed size again. I think the important thing here is that we check all the returned results.

mjsax commented on flink-contrib/flink-storm/src/test/java/org/apache/flink/storm/wrappers/BoltWrapperTest.java:
```diff
@@ -265,12 +264,12 @@ public void testOpen() throws Exception {
 		wrapper.open();
 		
-		verify(bolt).prepare(any(Map.class), any(TopologyContext.class), any(OutputCollector.class));
+		verify(bolt).prepare(any(Map.class), any(TopologyContext.class), isNotNull(OutputCollector.class));
```
Just out of curiosity: Why `isNotNull` instead of `any`?

mxm replied:
`any` matches also null (literally anything) but here I want to explicitly check `isNotNull`.

mjsax replied:
I see. Makes sense.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/WrapperSetupHelper.java:
```diff
@@ -224,7 +224,7 @@ static synchronized TopologyContext createTopologyContext(
 	 * 
 	 * @return A unique task ID if the currently processed Spout or Bolt ({@code componentId}) is equal to the current
-	 *         Flink operator ({@link operatorName}) -- {@code null} otherwise.
+	 *         Flink operator ({@param operatorName}) -- {@code null} otherwise.
```
I guess `@link` is wrong, but does `@param` work? Maybe `@code` would be correct? But I am not sure.

mxm replied:
Ok, let's use code.

http://stackoverflow.com/questions/1667212/reference-a-method-parameter-in-javadoc

mxm commented:
I've addressed your comments.

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -89,13 +99,9 @@ public BoltWrapper(final IRichBolt bolt) throws IllegalArgumentException {
+	 * @param bolt The Storm {@link IRichBolt bolt} to be used.
 	 * @param inputSchema
-	 *            The schema (ie, ordered field names) of the input stream.
-	 * @throws IllegalArgumentException
```
Please keep `@throws`

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/BoltWrapper.java:
```diff
@@ -108,16 +114,13 @@ public BoltWrapper(final IRichBolt bolt, final Fields inputSchema)
 	 * @throws IllegalArgumentException
 	 *             If {@code rawOuput} is {@code true} and the number of declared output attributes is not 1 or if
 	 *             {@code rawOuput} is {@code false} and the number of declared output attributes is not with range
-	 *             [1;25].
```
Please keep and update: should be `[0;25]`

mjsax commented on flink-contrib/flink-storm/src/main/java/org/apache/flink/storm/wrappers/StormTuple.java:
```diff
@@ -44,16 +47,32 @@
+	private final String producerStreamId;
+	/** The producer's component id of this tuple */
+	private final String producerComponentId;
+	/*+ The message that is associated with this tuple */
```
`/**` not `+`

mjsax commented:
If Travis is green please merge. You can fix the last tiny comments directly before merging. No need to update this PR.

mxm commented:
Thanks for your feedback!

mjsax commented:
Thanks for you patience. :)
